timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockWatchingServiceImpl.java

```diff
@@ -147,14 +156,33 @@ private void logLockWatchEvent(LockWatches newWatches) {
 
     private void runIfDescriptorsMatchLockWatches(
             Set<LockDescriptor> unfiltered, Consumer<Set<LockDescriptor>> consumer) {
+        runIfDescriptorsMatchLockWatchesWithMetadata(
+                unfiltered, Optional.empty(), (lockDescriptors, unused) -> consumer.accept(lockDescriptors));
+    }
+
+    private void runIfDescriptorsMatchLockWatchesWithMetadata(
+            Set<LockDescriptor> unfilteredLocks,
+            Optional<LockRequestMetadata> unfilteredMetadata,
+            BiConsumer<Set<LockDescriptor>, Optional<LockRequestMetadata>> biConsumer) {
         watchesLock.readLock().lock();
         try {
             RangeSet<LockDescriptor> ranges = watches.get().ranges();
-            Set<LockDescriptor> filtered =
-                    unfiltered.stream().filter(ranges::contains).collect(Collectors.toSet());
-            if (!filtered.isEmpty()) {
-                consumer.accept(filtered);
+            Set<LockDescriptor> filteredLocks =
+                    unfilteredLocks.stream().filter(ranges::contains).collect(Collectors.toSet());
+            if (filteredLocks.isEmpty()) {
+                return;
             }
+            // Filtering metadata after deciding if we should even proceed might save us some computation
+            Optional<LockRequestMetadata> filteredMetadata = unfilteredMetadata
+                    .map(LockRequestMetadata::lockDescriptorToChangeMetadata)
+                    .map(unfilteredLockMetadata -> {
+                        Map<LockDescriptor, ChangeMetadata> filteredLockMetadata = KeyedStream.ofEntries(
+                                        unfilteredLockMetadata.entrySet().stream())
+                                .filterKeys(ranges::contains)
```

Comment:
Might be worth adding a comment reasoning here why we can't just filter based upon the keys of the `filtered` LockDescriptors, and also why we are ok with simply matching on the lock watch ranges

Reply:
Interesting thought came up, which you may have already dealt with before, but: what happens if the Set of lock descriptors are disjoint from the map of lock descriptors w/metadata?

Reply:
Now that I think about it more, we will probably enforce that metadata can only be specified for lock descriptors that are part of the request/event. Otherwise, we will have problems with our wire format (index encoding). I think I will add this constraint checking when I implement client/server-side limits for the amounts of metadata. 

In that case, checking our filtered locks is definitely faster than doing `range.contains` (I did not think of this optimization, so good spot by you!).

This also answers your second comment. We will not allow disjoint sets, at least for `lockDescriptorToChangeMetadata`. If we ever had differently shaped metadata, we might want it anyways. This metadata will be a separate field in `LockRequestMetadata`, though, so it should not affect the filtering introduced in this PR.

Reply:
yeah not using `range.contains` is a bit of spicy change, but should be fine given that we've the invariants you mentioned :)

Reply:
How do you feel about splitting this up into it's own method? would be easier for us to reason about documentation.

Reply:
`filterMetadataToDescriptors` or something of the sort.